### DISCUSSION TITLE
Image Middleware: Transform Type/Format

### DISCUSSION THREAD

**Discussion by dimitrov-adrian** [directus/directus#3057] [Feature Requests]
In web apps it is good to display images in proper formats. As **webp** for example supports transparency and even animation, and since `sharp` support formatting the image at end, it will be good to have such option.

May be something as query parameter like
http://localhost:8080/assets/247f9f9d-8422-47e2-a537-d6b1abb25f76?key=system-medium-cover&type=webp that will works for images only as only images passes the transformations.

**Comment by philleepflorence** [comment]
> Cannot wait to drop `Imagemagick` for `Sharp`: 26.84 Operations per second for Files is amazing!

### Questions
- Can a user currently transform `JPEG to WEBp`, `PNG to JPEG`, or `PNG to WEBp` via the Assets Endpoint?
- How feasible would it be to allow [Assets Service](https://github.com/directus/directus/blob/main/api/src/services/assets.ts) Extensions?
    - <del>Create and Register `Hook` under `Assets` Scope or Override/Extend `Assets` Service.</del>
    - <del>_This could allow `Assets` to remain compact and agnostic but open up the many additional processing allowed by `Sharp` API_</del>

### Documentation and References
[Sharp: Directus Assets Image Processing Library](https://sharp.pixelplumbing.com)
[Thumbnail Transformations: Directus Concept](https://docs.directus.io/concepts/files/#thumbnail-transformations)
[Requesting a Thumbnail: Directus API Reference](https://docs.directus.io/reference/api/assets/#requesting-a-thumbnail)
[Custom API Hooks: Directus Guide](https://docs.directus.io/guides/api-hooks/)

**Reply by rijkvanzanten** [reply]
> Can a user currently transform JPEG to WEBp, PNG to JPEG, or PNG to WEBp via the Assets Endpoint?

See https://github.com/directus/directus/pull/3154#issuecomment-842372997 

> How feasible would it be to allow Assets Service Extensions?

We'll have to be very careful around image stream handling and transforming, as it's a very easy way to accidentally cause memory issues and file corruption

**Reply by philleepflorence** [reply]
> See #3154 (comment)

A JSON Dictionary matching Sharp [Resize](https://sharp.pixelplumbing.com/api-resize), [Output](https://sharp.pixelplumbing.com/api-output) Options actually seems simpler..._the more you know!_
Thanks for the PR link!

> We'll have to be very careful around image stream handling and transforming, as it's a very easy way to accidentally cause memory issues and file corruption

Oh, that I know very intimately, bypassed [D8 Thumbnailer](https://github.com/philleepflorence/directus-8-extensions/blob/master/helpers/filesystem.php) for many projects that had unique file and memory management requirements... _going the JSON options route seems to work for the use cases I can recollect_.
For `file corruption`, I am assuming the copy (stream), not the original, correct?

**Comment by philleepflorence** [comment]
## Feature Request Status
[Add format query parameter to assets: Pull Request](https://github.com/directus/directus/pull/3154)

**Comment by rijkvanzanten** [comment]
This now exists!
